### ISSUE TITLE
Swift support

### ISSUE THREAD

**Issue by aleksrutins** [zed-industries/zed#5278]
**Is your feature request related to a problem? Please describe.**
I'd like to be able to use Zed to work on my Swift projects.

**Describe the solution you'd like**
[SourceKit-LSP](https://github.com/apple/sourcekit-lsp) could be used.

**Describe alternatives you've considered**
N/A?

**Screenshots**
N/A

**Comment by marcusziade** [comment]
Any updates?

**Comment by marcusziade** [comment]
If someone can explain what the starting point of adding Swift support to this is, I can attempt to do it

**Comment by ValentinWalter** [comment]
> If someone can explain what the starting point of adding Swift support to this is, I can attempt to do it

You'd need to integrate with the Swift LSP. From a quick glance at the repo it looks like https://github.com/zed-industries/zed/blob/main/plugins/json_language/src/lib.rs might be a good place to start.

**Comment by fdionisi** [comment]
I think the plugin story may take a bit longer, but it's part of the [roadmap](https://zed.dev/roadmap). You may want to wait a little longer or join the conversation in Zed, otherwise take a look at how other languages are implemented in [`language.rs`](https://github.com/zed-industries/zed/blob/90c1d8f734615552297a65e12dd430bf0825cdb0/crates/zed/src/languages.rs).

Also, https://github.com/zed-industries/zed/pull/5816 is a good one on how to add an LSP for a language.

**Comment by iamnbutler** [comment]
I tried adding swift a while back – the issue I ran into was the treesitter grammar was so immensely slow it could take 10-30 seconds on a 32core 64gb ram m1 for a single file to highlight – I ended up abandoning it due to that experience being awful.

**Comment by iamnbutler** [comment]
I got it working in a few hours so I'm sure someone can reproduce it pretty quickly to see if the experience has improved – if it hasn't we would likely need to wait until there is a better grammar, I don't think we will be willing to ship a language that impacts editor perf to that degree.

**Comment by iamnbutler** [comment]
Feel free to pick up from here: https://github.com/zed-industries/zed/pull/2748. This is Zed1 though so the approach may be different, I'm not sure.

**Comment by fbartho** [comment]
@iamnbutler silly question: why did you go with tree-sitter-swift instead of building on the official LSP?

Would the official LSP have been slower? Or was it more that tree-sitter was already a common pattern so it was less work?

**Comment by marcusziade** [comment]
I've integrated the sourcekit-lsp into Neovim and had no performance issues

**Comment by iamnbutler** [comment]
> @iamnbutler silly question: why did you go with tree-sitter-swift instead of building on the official LSP?
> 
> 
> 
> Would the official LSP have been slower? Or was it more that tree-sitter was already a common pattern so it was less work?

Forgive me if I'm wrong, but you will still need a tree-sitter grammar to get syntax highlighting right?

**Comment by Saafo** [comment]
can we build the treesitter above nvim-treesitter?

**Comment by ivanglushko** [comment]
> If someone can explain what the starting point of adding Swift support to this is, I can attempt to do it

Awesome! Just curious, have you started yet?

**Comment by marcusziade** [comment]
> > If someone can explain what the starting point of adding Swift support to this is, I can attempt to do it
> 
> Awesome! Just curious, have you started yet?

No, I haven't started and got turned off by the comments above. I'm also busy with my side project and don't wanna take on too much at this time.

**Comment by ceopaludetto** [comment]
I've started to play around with zed codebase and tree-sitter-swift but I quickly found this issue https://github.com/alex-pinkus/tree-sitter-swift/issues/240, some queries are really slow. Anyway, it kinda works.

<img width="2672" alt="Screenshot 2024-01-26 at 16 06 42" src="https://github.com/zed-industries/zed/assets/30508218/dfe40847-08a2-42a0-8602-aab540aff8ae">

**Comment by marcusziade** [comment]
Would you give an example of what a "slow query" actually looks like while using?

**Comment by ceopaludetto** [comment]
> Would you demonstrate what a "slow query" actually looks like in practice?
> 
> Wouldn't surprise me if the slowest slowdowns in Zed are faster than the fastest things in Xcode.

I've added the suggested `highlights.scm` file which you can found here https://github.com/alex-pinkus/tree-sitter-swift/blob/main/queries/highlights.scm. Even on a small file(15 lines) it took approximately 30 seconds to highlight on my M1 Pro in debug mode, example video:

https://github.com/zed-industries/zed/assets/30508218/32772337-d03e-4e82-8357-5d632b0e8c52

Also the LSP does not work(and only start to process after tree-sitter parsing mechanism). I'll draft a pull request because I'm probably doing something wrong.

**Comment by marcusziade** [comment]
Is it only an initial load or for every file change? Do Git diffs affect it somehow?

**Comment by ceopaludetto** [comment]
> Is it only an initial load or for every file change? Do Git diffs affect it somehow?

On initial load only, subsequent modifications work as expected. I also tested on a package without `.git` and it does not seem related.

Draft PR: https://github.com/zed-industries/zed/pull/6831

**Comment by marcusziade** [comment]
If it's only once, I wouldn't mind that much. The benefits outweigh that.

**Comment by ADGrant** [comment]
This is a very interesting project and I like the focus on performance. I think it's unfortunate that an editor only available on MacOS doesn't support Swift though. Most Swift developers do use Macs.

**Comment by iamnbutler** [comment]
Yeah those slowdowns were the same reason I abandoned my PR last year - from our PoV in an editor all about performance it feels pretty bad to have a supported language take 30 seconds to highlight. 

Now that extensions are coming, if someone wanted to opt in to that that would be fine.

I do wonder what the root cause of all the slowness is though.

**Comment by marcusziade** [comment]
Is it using Swift-syntax? That has caused many slowdowns for many

**Comment by subdigital** [comment]
I’ll admit that sourcekit-lsp is pretty limited, but performance with completion and highlighting with tree sitter are not an issue in neovim.Sent from my iPhoneOn Feb 14, 2024, at 3:11 PM, Marcus Ziadé ***@***.***> wrote:﻿
Is it using Swift-syntax? That has caused many slowdowns for many

**Comment by marcusziade** [comment]
I'm using sourcekit-lsp in Neovim as well and see no slowdowns.

**Comment by pblazej** [comment]
Used by [Swift VSCode](https://marketplace.visualstudio.com/items?itemName=sswg.swift-lang) as well, works fine.

**Comment by ostwilkens** [comment]
It's not uncommon for my XCode to randomly have 1s input lag. Being able to code in Zed would be a breath of fresh air.

**Comment by megabitsenmzq** [comment]
If we also use [InjectionIII](https://github.com/johnno1962/InjectionIII), we can even have a real-time preview alongside Zed. 😂
